### PR TITLE
feat: Improved setConfiguration command with automatic configuration detection

### PR DESCRIPTION
**Pull request by Vigilans**
## Demonstration

image
image
image

### CODE REVIEW

File: src/constants/checkstyleConfigs.ts [hunk 1]
```diff
@@ -5,3 +5,14 @@ export enum BuiltinConfiguration {
     GoogleCheck = '/google_checks.xml',
     SunCheck = '/sun_checks.xml',
 }
+
+export const DoctypePublicIds: string[] = [
```
Comment (jdneo) [inline src/constants/checkstyleConfigs.ts]: rename to `checkstyleDocTypeIds`?

File: src/utils/settingUtils.ts [hunk 2]
```diff
@@ -40,6 +41,15 @@ export function isAutoCheckEnabled(): boolean {
     return getConfiguration().get<boolean>(JAVA_CHECKSTYLE_AUTOCHECK, true);
 }
 
+export function asWorkspaceEnvPath(fsPath: string): string {
```
Comment (jdneo) [inline src/utils/settingUtils.ts]: `useWorkspaceFolderIfPossible`

File: src/commands/config.ts [hunk 3]
```diff
@@ -1,85 +1,105 @@
...
+import * as xmljs from 'xml-js';
+import { BuiltinConfiguration, DoctypePublicIds } from '../constants/checkstyleConfigs';
+import { asWorkspaceEnvPath as tryAsWorkspaceEnvPath, getDefaultWorkspaceFolder, setCheckstyleConfigurationPath } from '../utils/settingUtils';
 
-export async function setCheckstyleConfiguration(uri?: Uri): Promise<void> {
+export async function setConfiguration(uri?: Uri): Promise<void> {
     if (uri) {
         if (path.extname(uri.fsPath).toLowerCase() === '.xml') {
-            setCheckstyleConfigurationPath(uri.fsPath, uri);
+            setCheckstyleConfigurationPath(tryAsWorkspaceEnvPath(uri.fsPath), uri);
```
Comment (jdneo) [inline src/commands/config.ts]: What will happen if it's not a checkstyle xml?

File: src/commands/config.ts [hunk 4]
```diff
@@ -1,85 +1,105 @@
...
+        const relativeXml: string = tryAsWorkspaceEnvPath(xml.fsPath);
+        function doctypeFn(doctype: string): void {
+            const [name, type] = doctype.split(/\s+/, 2);
+            if (type.toUpperCase() === 'PUBLIC') {
+                const pubid: string = doctype.match(/"(.+)"/)![1];
+                if (DoctypePublicIds.includes(pubid)) {
+                    detected.push({ label: relativeXml, detail: xml.fsPath });
+                }
+            }
+            if (type.toUpperCase() === 'SYSTEM') {
```
Comment (jdneo) [inline src/commands/config.ts]: else if

File: src/commands/config.ts [hunk 5]
```diff
@@ -1,85 +1,104 @@
...
         {
-            label: ConfigurationSelction.SunStyle,
+            label: '$(file-text) Browse...',
+            detail: 'Select a configuration file from your file system.',
         },
     ];
     const result: QuickPickItem | undefined = await window.showQuickPick(items, { ignoreFocusOut: true });
-    if (result) {
-        return result.label;
+    if (result === items.slice(-1)[0]) {
```
Comment (jdneo) [inline src/commands/config.ts]: extend the quickpickitem to have the value field and compare with it?

File: src/commands/config.ts (hunk 3 above)
Comment (Vigilans) [reply src/commands/config.ts]: It only handles the path of xml file, the behavior is consistent with previous versions if the uri path isn't in workspace folder.

**Pull request merged**